lib/locality.c

```diff
@@ -146,11 +146,17 @@ M0_INTERNAL void m0_locality_fini(struct m0_locality *loc)
 M0_INTERNAL struct m0_locality *m0_locality_here(void)
 {
 	struct locality_global *glob = loc_glob();
+	struct m0_locality     *loc;
 
 	if (glob->lg_dom == NULL || m0_thread_self() == &glob->lg_ast_thread)
 		return &glob->lg_fallback;
-	else
-		return m0_locality_get(m0_processor_id_get());
+	else {
+		loc = m0_thread_tls()->tls_loc;
+		if (loc == NULL) /* unlikely, see loc_handler_thread() */
+			loc = m0_thread_tls()->tls_loc =
+				m0_locality_get(m0_processor_id_get());
```

Comment:
This is still not completely bullet-proof: if locality handler thread migrates to a different core before `m0_thread_tls()->tls_loc = &loc->fl_locality;` assignment happens, things will go wrong.

I prefer a more systematic approach: assign ->tls_loc to the current processor in m0_thread_init(). In m0_thread_confine(), reset it to the first processor in the bitmap.

Reply:
Ok, updated.

Reply:
BTW, what about threads which do not call m0_thread_init()? What if they call m0_locality_here()?

Reply:
I think I know the answer now - there should no be such threads. Only Motr threads may call `m0_locality_here()`, and all Motr threads are created via `M0_THREAD_INIT()` -> `m0_thread_init()`.